applications/DEMApplication/python_scripts/DEM_analysis_stage.py

```diff
@@ -1072,41 +1072,110 @@ def MeasureSphereForGettingPackingProperties(self, radius, center_x, center_y, c
                     center_to_sphere_distance_1 = ((x_1 - center_x)**2 + (y_1 - center_y)**2 + (z_1 - center_z)**2)**0.5
 
-                    if center_to_sphere_distance_0 < (radius - r_0):
+                    if center_to_sphere_distance_0 < (radius - r_0) or center_to_sphere_distance_1 < (radius - r_1):
+                        d = ((x_0 - x_1)**2 + (y_0 - y_1)**2 + (z_0 - z_1)**2)**0.5
+                        if d <= (r_0 + r_1):
+                            vector1 = np.array([x_1 - x_0 , y_1 - y_0, z_1 - z_0])
+                            v1_norm = np.linalg.norm(vector1)
+                            if v1_norm:
+                                vector1_unit = vector1 / v1_norm
+                            tensor = np.outer(vector1_unit, vector1_unit)
+                            total_tensor += tensor
+                            total_contact_number += 1
+                        else:
+                            continue
 
-                        vector1 = np.array([x_1 - x_0 , y_1 - y_0, z_1 - z_0])
-                        v1_norm = np.linalg.norm(vector1)
-                        if v1_norm:
-                            vector1_unit = vector1 / v1_norm
-                        tensor = np.outer(vector1_unit, vector1_unit)
-                        total_tensor += tensor
-                        total_contact_number += 1
-
-                    elif center_to_sphere_distance_1 < (radius - r_1):
-
-                        vector1 = np.array([x_1 - x_0 , y_1 - y_0, z_1 - z_0])
-                        v1_norm = np.linalg.norm(vector1)
-                        if v1_norm:
-                            vector1_unit = vector1 / v1_norm
-                        tensor = np.outer(vector1_unit, vector1_unit)
-                        total_tensor += tensor
-                        total_contact_number += 1
-                
                 if total_contact_number:
                     measured_fabric_tensor = total_tensor / total_contact_number
+
                 else:
                     measured_fabric_tensor = np.empty((3, 3))
 
-                deviatoric_tensor = 4 * (measured_fabric_tensor - 1/3 * np.eye(3)) 
+                deviatoric_tensor = 4 * (measured_fabric_tensor - 1/3 * np.eye(3))
 
                 second_invariant_of_deviatoric_tensor = (0.5 * np.sum(deviatoric_tensor * deviatoric_tensor))**0.5
 
+                fabric_tensor_trace = (measured_fabric_tensor[0][0] + measured_fabric_tensor[1][1] + measured_fabric_tensor[2][2])/3
+
                 eigenvalues, eigenvectors = np.linalg.eig(measured_fabric_tensor)
-                
-                return eigenvalues, second_invariant_of_deviatoric_tensor, measured_fabric_tensor
+
+                return eigenvalues, second_invariant_of_deviatoric_tensor, [measured_fabric_tensor[0][0],measured_fabric_tensor[1][1],measured_fabric_tensor[2][2]], fabric_tensor_trace
```

Comment:
Why return "[measured_fabric_tensor[0][0],measured_fabric_tensor[1][1],measured_fabric_tensor[2][2]]" instead of "measured_fabric_tensor"? I believe the "measured_fabric_tensor" includes the information you need. But if the output is "[measured_fabric_tensor[0][0],measured_fabric_tensor[1][1],measured_fabric_tensor[2][2]]", some cases may be broken.

Reply:
Because I need the diagonal elements of the tensor, the measured fabric tensor variable you wrote in the past only gives the trace of the tensor.

Reply:
Yes, I can see that you need the diagonal elements of the tensor. What I mean is that you can get those diagonal elements by using "[measured_fabric_tensor[0][0],measured_fabric_tensor[1][1],measured_fabric_tensor[2][2]]" in a specific place. but we cannot do the opposite (from "[measured_fabric_tensor[0][0],measured_fabric_tensor[1][1],measured_fabric_tensor[2][2]]" get the "measured_fabric_tensor"). The cases I mentioned are not in tests now. I believe "measured_fabric_tensor" is more general.